### ISSUE TITLE
offline drm widevine for hls stream-How to extract drminitdata?

### ISSUE THREAD

**Issue by fridapinto-debug** [google/ExoPlayer#8240]
for dash video offline license key can be done like this:
```

                DataSource dataSource = httpDataSourceFactory.createDataSource();
                DashManifest dashManifest = DashUtil.loadManifest(dataSource, getIntent().getData());
                DrmInitData drmInitData = DashUtil.loadDrmInitData(dataSource, dashManifest.getPeriod(0));
```

But I want to know how to extract the drminitData for hls stream with widevine protection?

**Comment by AquilesCanta** [comment]
Hi @fridapinto-debug, there's no `DashUtil` equivalent for HLS and you should implement the fetching yourself. I don't discard that an `HlsUtil` could exist in the future, but it's actually better if you build it yourself, since (a) it's just a handful of lines and (b) the process depends a lot on the shape of your media and your usecase.

What you need to do is replace DashUtil's `DashManifestParser` with an `HlsPlaylistParser`, then cast the result of parsing (which is an `HlsPlaylist`) to whatever it is you are loading (could be a [master](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/source/hls/playlist/HlsMasterPlaylist.html) or [media](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/source/hls/playlist/HlsMediaPlaylist.html) playlist).

If it's a master playlist, then if you are using a [session key](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/source/hls/playlist/HlsMasterPlaylist.html#sessionKeyDrmInitData), you might want to use that as DrmInitData. If you are not using a session key, you'll need to load at least one of the media playlists (see [variants](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/source/hls/playlist/HlsMasterPlaylist.html#variants) and [renditions](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/source/hls/playlist/HlsMasterPlaylist.html#audios) - Usually only audio). You might want to download more than one if each media playlist uses different DRM keys (for instance, video and audio are demuxed and their keys are served separately).

Once you have downloaded a MediaPlaylist, you need to know how your streams are protected. You'll have to iterate over the segments and get the DrmInitDatas (Use [HlsMediaPlaylist.Segment.drmInitData](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/source/hls/playlist/HlsMediaPlaylist.Segment.html#drmInitData)) for which you want to download the keys. If your stream uses the same key for the entire duration, the first DrmInitData will be what you are looking for. But if your stream does key rotation, then you'll need to get all distinct DrmInitData's in the playlist. As mentioned before, if each media playlist uses different keys, you'll need to iterate over the segments of each media playlist you want to download (usually only one, though).

In conclusion, this will be a bit more complicated than Dash, but since you probably know the shape of the media better than us, it will probably be much simpler for you and you'll be able to more efficiently fetch the key acquisition data. Please let me know if this works for you so we can close the issue.
